D3ASIM-1350: Tracing origin of ESS Energy

faizan2590 opened the pull request.

hannesdiedrich commented on src/d3a/d3a_core/export.py:
```diff
@@ -436,6 +438,55 @@ def _plot_ess_soc_history(self, storage_list: list, subdir: str, root_name: str)
+            temp[ESSEnergyOrigin.EXTERNAL][k] = v[ESSEnergyOrigin.EXTERNAL]
+            temp[ESSEnergyOrigin.LOCAL][k] = v[ESSEnergyOrigin.LOCAL]
+            temp[ESSEnergyOrigin.UNKNOWN][k] = v[ESSEnergyOrigin.UNKNOWN]
+        for i in [ESSEnergyOrigin.EXTERNAL, ESSEnergyOrigin.LOCAL, ESSEnergyOrigin.UNKNOWN]:
```
Please omit one character variables for better readability. Thanks!

hannesdiedrich commented on src/d3a/d3a_core/export.py:
```diff
@@ -436,6 +438,55 @@ def _plot_ess_soc_history(self, storage_list: list, subdir: str, root_name: str)
+                ESSEnergyOrigin.LOCAL: {slot: 0. for slot in generate_market_slot_list()},
+                ESSEnergyOrigin.EXTERNAL: {slot: 0. for slot in generate_market_slot_list()}}
+
+        for k, v in graph_obj.dataset.items():
```
Can you please explain why this step is necessary ? Also, why are you initialising the PlotlyGraph and also go.Bar object? I think that PlotlyGraph.plot_bar_graph() does everything you want. You just have to put the data in the right format.

faizan2590 replied:
Data is in different format due to which PlotlyGraph.plot_bar_graph() wouldn't be able to work.

hannesdiedrich commented on src/d3a/models/state.py:
```diff
@@ -107,6 +123,8 @@ def __init__(self,
 
         self._used_storage = initial_capacity_kWh
         self._battery_energy_per_slot = 0.0
+        # list of namedtuple
```
You can remove this comment. Thanks

hannesdiedrich commented on src/d3a/models/strategy/storage.py:
```diff
@@ -226,15 +231,45 @@ def event_tick(self, *, area):
+            if child.name == trade.seller:
+                return True
+
+    def _energy_sell_type(self, trade):
```
As this is only for tracking, please rename e.g. to `_track_energy_sell_type`

hannesdiedrich replied:
The same for `_energy_bought_type`

hannesdiedrich replied:
Also, please leave a comment here, that the FIRST-IN FIRST-OUT mechanism is used (just for later). Thanks!

hannesdiedrich commented on src/d3a/models/strategy/storage.py:
```diff
@@ -226,15 +231,45 @@ def event_tick(self, *, area):
+    def _energy_sell_type(self, trade):
+        energy = trade.offer.energy
+        while limit_float_precision(energy) > 0:
+            recent_energy = self.state.get_used_storage_share[0].value
```
I could think of a renaming of `recent_energy` to `first_in_energy`. What do you think about that?

hannesdiedrich commented on src/d3a/models/strategy/storage.py:
```diff
@@ -226,15 +231,45 @@ def event_tick(self, *, area):
+                energy -= recent_energy
+                self.state.get_used_storage_share.pop(0)
+            elif energy < recent_energy:
+                last = self.state.get_used_storage_share[0]
```
In order to avoid confusion, I suggest to omit the declaration of `last`  and put `self.state.get_used_storage_share[0]` into line 256. Does it make sense?

faizan2590 replied:
Please have a look at it here: https://github.com/gridsingularity/d3a/pull/497/files#diff-6479cb83f0833c9660b171d3fd0bfc57R250

hannesdiedrich commented on tests/test_strategy_storage.py:
```diff
@@ -831,3 +833,66 @@ def test_initial_selling_rate(storage_strategy_test14, area_test14):
+        Trade('id', 'time', Offer('id', 20, 1.0, 'ChildArea'), 'ChildArea', 'FakeArea')
+    storage_strategy_test15.event_trade(market_id=market_test15.id,
+                                        trade=storage_strategy_test15.area.current_market.trade)
+    storage_strategy_test15.event_market_cycle()
```
For this test you dont need to call event_market_cycle(), right? Only if you want to test whether state.time_series_ess_share is written. Or did i get it wrong?

faizan2590 replied:
Yes, you are right. Removing it

hannesdiedrich commented on src/d3a/d3a_core/export.py:
```diff
@@ -436,6 +438,55 @@ def _plot_ess_soc_history(self, storage_list: list, subdir: str, root_name: str)
+        """
+
+        new_subdir = os.path.join(subdir, area.slug)
+        from d3a.models.strategy.storage import StorageStrategy
```
Please move to the top. Thanks!

faizan2590 commented:
> Generally, the code makes sense.
> I left some comments for refactoring and a few questions.
> Once they are dealt with, i am happy to review again.
> Additionally, i think it is worth it to add an integrationtest for testing the whole workflow.

I have tried to addressed your comments. Please have a look. Considering integration_test, lets discuss it offline
